### PR TITLE
JAMES-3946 add a memory implementation of the DropList

### PR DESCRIPTION
**Pull request by Maxxx873**
[https://issues.apache.org/jira/projects/JAMES/issues/JAMES-3946](https://issues.apache.org/jira/projects/JAMES/issues/JAMES-3946?filter=allopenissues)

### CODE REVIEW

File: server/data/data-api/src/main/java/org/apache/james/droplists/api/DropListEntry.java [hunk 1]
```diff
@@ -71,6 +76,11 @@ public DropListEntry build() {
             Preconditions.checkArgument(deniedEntity != null && !deniedEntity.isBlank(), "`deniedEntity` must not be null, empty, or blank");
             Preconditions.checkArgument(deniedEntity.length() <= MAXIMUM_DOMAIN_LENGTH,
                 "deniedEntity length should not be longer than %s characters", MAXIMUM_DOMAIN_LENGTH);
+            if (deniedEntityType.equals(DeniedEntityType.DOMAIN)) {
+                deniedEntity = Domain.of(deniedEntity).asString();
+            } else {
+                deniedEntity = new MailAddress(deniedEntity).toString();
+            }
```
Comment (chibenwa) [inline server/data/data-api/src/main/java/org/apache/james/droplists/api/DropListEntry.java:83]: Generally speaking we can avoid variable reallocation by method extraction.
This is more elegant and less complex.

**Review by chibenwa — COMMENTED** [review]
Hello,

Thanks to put this together!

While the datastructure for the memory structure is not optimal (`O(N)` scans)
And while we could easily do better
(one multimap for global scope, one table for domain scope, one table for user scope) this does not matter much performance wise in a testing environment.

Other remarks however likely deserve attention.

**Comment by Maxxx873** [comment]
> Hello,
> 
> Thanks to put this together!
> 
> While the datastructure for the memory structure is not optimal (`O(N)` scans) And while we could easily do better (one multimap for global scope, one table for domain scope, one table for user scope) this does not matter much performance wise in a testing environment.
> 
> Other remarks however likely deserve attention.

Thanks for the excellent code review! I will try to correct everything that was pointed out to me in the remarks.

File: server/data/data-api/src/test/java/org/apache/james/droplists/api/DropListContract.java [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+
+public interface DropListContract {
+
+    DropList dropList();
+
+    @Test
+    default void shouldAddEntry() throws AddressException {
+        DropListEntry dropListEntry = getDropListTestEntries().toList().getFirst();
```
Comment (chibenwa) [inline server/data/data-api/src/test/java/org/apache/james/droplists/api/DropListContract.java:45]: That's quite complex for just getting one item.

Calling the builder here without hiding things behind a method call would seem more straight forward to me...

File: server/data/data-api/src/test/java/org/apache/james/droplists/api/DropListContract.java [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+
+        Mono<Void> result = dropList().add(dropListEntry);
+
+        assertThat(Objects.requireNonNull(dropList().list(GLOBAL, dropListEntry.getOwner()).collectList().block()).size()).isEqualTo(1);
```
Comment (chibenwa) [inline server/data/data-api/src/test/java/org/apache/james/droplists/api/DropListContract.java:49]: I think Objects.requireNonNull is overkill here.

 -> A NPE thrown would still be understandable for a dev debugging this
 -> Reading is harder and we do not have such a convention at all through the code base.
 
 Suggestion: relax your IDE warning or ignore those.

**Review by chibenwa — COMMENTED** [review]
Looks good to me. Tiny cosmetic fixes.

Can you confirm the test suite establishes a complete test coverage?

**Review by Arsnael — COMMENTED** [review]
Read it, nothing to add.

Thank you for the work, appreciated as always :)

**Comment by Maxxx873** [comment]
> Read it, nothing to add.
> 
> Thank you for the work, appreciated as always :)

Thank you too, I'm really glad to be involved in this project and follow its development.

**Comment by Maxxx873** [comment]
> Looks good to me. Tiny cosmetic fixes.
> 
> Can you confirm the test suite establishes a complete test coverage?

image

**Review by Arsnael — APPROVED** [review]

**Review by chibenwa — APPROVED** [review]

**Pull request merged**